File: pkg/parser/parser.go
package parser

import (
	"encoding/binary"
	"encoding/json"
	"encoding/xml"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/agent-cairn/criu-inspector/pkg/models"
)

// ParseCheckpoint parses a CRIU checkpoint directory
func ParseCheckpoint(dir string, verbose bool) (*models.CheckpointData, error) {
	// Validate directory exists
	if _, err := os.Stat(dir); os.IsNotExist(err) {
		return nil, fmt.Errorf("checkpoint directory not found: %s", dir)
	}
	data := &models.CheckpointData{}

	// Parse pstree.img (always)
	pstreePath := filepath.Join(dir, "pstree.img")
	if _, err := os.Stat(pstreePath); err == nil {
		processes, err := ParsePstree(pstreePath)
		if err != nil {
			return nil, fmt.Errorf("failed to parse pstree.img: %w", err)
		}
		data.Processes = processes
	}

	// Parse inventory.img (verbose only)
	if verbose {
		inventoryPath := filepath.Join(dir, "inventory.img")
		if _, err := os.Stat(inventoryPath); err == nil {
			inventory, err := ParseInventory(inventoryPath)
			if err != nil {
				return nil, fmt.Errorf("failed to parse inventory.img: %w", err)
			}
			data.Inventory = inventory
		}
	}

	// Parse fds.img (verbose only)
	if verbose {
		fdsPath := filepath.Join(dir, "fds.img")
		if _, err := os.Stat(fdsPath); err == nil {
			fds, err := ParseFds(fdsPath)
			if err != nil {
				return nil, fmt.Errorf("failed to parse fds.img: %w", err)
			}
			data.FileDescriptors = fds
		}
	}

	return data, nil
}

// ParsePstree parses the pstree.img file (binary format with embedded JSON)
func ParsePstree(path string) ([]*models.ProcessEntry, error) {
	content, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("failed to read pstree.img: %w", err)
	}

	// CRIU pstree.img has a binary header, JSON starts after '['
	// Search for the first '[' character
	jsonStart := -1
	for i, b := range content {
		if b == '[' {
			jsonStart = i
			break
		}
	}

	if jsonStart == -1 {
		// No JSON found, try binary parsing
		entries, err := parsePstreeBinary(content)
		if err != nil {
			return nil, fmt.Errorf("failed to parse pstree: no JSON found and binary parse failed: %w", err)
		}
		return entries, nil
	}

	// Parse JSON portion
	jsonBytes := content[jsonStart:]
	var entries []*models.ProcessEntry
	if err := json.Unmarshal(jsonBytes, &entries); err != nil {
		return nil, fmt.Errorf("failed to unmarshal JSON: %w", err)
	}

	return entries, nil
}

// parsePstreeBinary attempts to parse pstree.img as a binary format
// This is a fallback for when no JSON is embedded
func parsePstreeBinary(content []byte) ([]*models.ProcessEntry, error) {
	// This is a simplified parser for the binary format
	// Real CRIU uses a complex protobuf-like format
	// For demo purposes, we'll extract what we can

	entries := []*models.ProcessEntry{}

	// Skip header (first 100 bytes typically)
	if len(content) < 100 {
		return nil, fmt.Errorf("content too short to be valid pstree.img")
	}

	// Try to find patterns that look like PID/PPID pairs
	// This is heuristic and may not work for all checkpoint formats
	offset := 100
	for offset < len(content)-16 {
		// Look for reasonable PID values (1-65535)
		pid := binary.LittleEndian.Uint32(content[offset : offset+4])
		ppid := binary.LittleEndian.Uint32(content[offset+4 : offset+8])

		// Heuristic: valid process tree entries
		if pid > 0 && pid < 65536 && ppid < 65536 {
			entry := &models.ProcessEntry{
				PID:      pid,
				PPID:     ppid,
				PGID:     pid, // Assume process group = PID
				SID:      pid, // Assume session = PID
				Comm:     "process", // Unknown without full parsing
				NThreads: 1,
			}
			entries = append(entries, entry)
			offset += 16 // Skip to next potential entry
		} else {
			offset++
		}
	}

	if len(entries) == 0 {
		return nil, fmt.Errorf("no valid process entries found in binary format")
	}

	return entries, nil
}

// ParseInventory parses inventory.img (XML format)
func ParseInventory(path string) (*models.Inventory, error) {
	content, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("failed to read inventory.img: %w", err)
	}

	// Try to parse as XML
	var inventory models.Inventory
	if err := xml.Unmarshal(content, &inventory); err != nil {
		// Fallback: extract key-value pairs manually
		return parseInventoryFallback(string(content))
	}

	return &inventory, nil
}

// parseInventoryFallback extracts inventory data manually if XML parsing fails
func parseInventoryFallback(content string) (*models.Inventory, error) {
	inventory := &models.Inventory{
		Magic:   "UNKNOWN",
		Arch:    "UNKNOWN",
		PagesID: "UNKNOWN",
		GenID:   "UNKNOWN",
	}

	lines := strings.Split(content, "\n")
	for _, line := range lines {
		line = strings.TrimSpace(line)
		if strings.Contains(line, "<magic>") {
			start := strings.Index(line, "<magic>") + 7
			end := strings.Index(line, "</magic>")
			if end > start {
				inventory.Magic = line[start:end]
			}
		} else if strings.Contains(line, "<arch>") {
			start := strings.Index(line, "<arch>") + 6
			end := strings.Index(line, "</arch>")
			if end > start {
				inventory.Arch = line[start:end]
			}
		} else if strings.Contains(line, "<pages_id>") {
			start := strings.Index(line, "<pages_id>") + 10
			end := strings.Index(line, "</pages_id>")
			if end > start {
				inventory.PagesID = line[start:end]
			}
		} else if strings.Contains(line, "<gen_id>") {
			start := strings.Index(line, "<gen_id>") + 8
			end := strings.Index(line, "</gen_id>")
			if end > start {
				inventory.GenID = line[start:end]
			}
		} else if strings.Contains(line, "<entries>") {
			start := strings.Index(line, "<entries>") + 9
			end := strings.Index(line, "</entries>")
			if end > start {
				var entries uint32
				fmt.Sscanf(line[start:end], "%d", &entries)
				inventory.Entries = entries
			}
		} else if strings.Contains(line, "<image_pages>") {
			start := strings.Index(line, "<image_pages>") + 12
			end := strings.Index(line, "</image_pages>")
			if end > start {
				var pages uint32
				fmt.Sscanf(line[start:end], "%d", &pages)
				inventory.ImagePages = pages
			}
		}
	}

	return inventory, nil
}

// ParseFds parses fds.img (XML format)
func ParseFds(path string) ([]*models.FileDescriptor, error) {
	content, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("failed to read fds.img: %w", err)
	}

	// Parse XML structure manually for robustness
	fds := []*models.FileDescriptor{}
	lines := strings.Split(string(content), "\n")

	var currentFD *models.FileDescriptor
	for _, line := range lines {
		line = strings.TrimSpace(line)

		// Start of FD element
		if strings.Contains(line, "<fd ") {
			currentFD = &models.FileDescriptor{
				Type: "unknown",
			}

			// Parse attributes
			if idx := strings.Index(line, `fd="`); idx >= 0 {
				start := idx + 4
				end := strings.Index(line[start:], `"`)
				if end > 0 {
					var fd int32
					fmt.Sscanf(line[start:start+end], "%d", &fd)
					currentFD.FD = fd
				}
			}

			if idx := strings.Index(line, `type="`); idx >= 0 {
				start := idx + 6
				end := strings.Index(line[start:], `"`)
				if end > 0 {
					currentFD.Type = line[start : start+end]
				}
			}

			if idx := strings.Index(line, `path="`); idx >= 0 {
				start := idx + 6
				end := strings.Index(line[start:], `"`)
				if end > 0 {
					currentFD.Path = line[start : start+end]
				}
			}

			if idx := strings.Index(line, `flags="`); idx >= 0 {
				start := idx + 7
				end := strings.Index(line[start:], `"`)
				if end > 0 {
					currentFD.Flags = line[start : start+end]
				}
			}

			if idx := strings.Index(line, `id="`); idx >= 0 {
				start := idx + 4
				end := strings.Index(line[start:], `"`)
				if end > 0 {
					currentFD.ID = line[start : start+end]
				}
			}

			if idx := strings.Index(line, `mnt_id="`); idx >= 0 {
				start := idx + 8
				end := strings.Index(line[start:], `"`)
				if end > 0 {
					var mntID uint32
					fmt.Sscanf(line[start:start+end], "%d", &mntID)
					currentFD.MntID = mntID
				}
			}

			// Self-closing tag
			if strings.Contains(line, "/>") {
				fds = append(fds, currentFD)
				currentFD = nil
			}
		} else if strings.Contains(line, "</fd>") && currentFD != nil {
			fds = append(fds, currentFD)
			currentFD = nil
		}
	}

	return fds, nil
}

File: src/parser.rs
use crate::error::{InspectorError, Result};
use crate::models::{CheckpointData, FileDescriptor, Inventory, ProcessEntry};
use std::fs;
use std::path::Path;

/// Parser for CRIU checkpoint files
pub struct CriuParser;

impl CriuParser {
    /// Parse a CRIU checkpoint directory
    pub fn parse_checkpoint_dir(dir: &Path, verbose: bool) -> Result<CheckpointData> {
        let mut data = CheckpointData::default();

        // Parse pstree.img (binary format with embedded JSON)
        let pstree_path = dir.join("pstree.img");
        if pstree_path.exists() {
            data.processes = Self::parse_pstree(&pstree_path)?;
        }

        // Parse inventory.img (XML format) - only in verbose mode
        if verbose {
            let inventory_path = dir.join("inventory.img");
            if inventory_path.exists() {
                data.inventory = Some(Self::parse_inventory(&inventory_path)?);
            }
        }

        // Parse fds.img (XML format) - only in verbose mode
        if verbose {
            let fds_path = dir.join("fds.img");
            if fds_path.exists() {
                data.file_descriptors = Self::parse_fds(&fds_path)?;
            }
        }

        Ok(data)
    }

    /// Parse pstree.img - extracts JSON from binary CRIU format
    fn parse_pstree(path: &Path) -> Result<Vec<ProcessEntry>> {
        let content = fs::read(path)?;

        // CRIU pstree.img format: binary header + JSON data
        // We'll search for the first '[' character to find JSON start
        let json_start = content
            .iter()
            .position(|&b| b == b'[')
            .ok_or_else(|| InspectorError::Parse("No JSON found in pstree.img".to_string()))?;

        let json_str = std::str::from_utf8(&content[json_start..])
            .map_err(|e| InspectorError::Parse(format!("Invalid UTF-8: {}", e)))?;

        let entries: Vec<ProcessEntry> = serde_json::from_str(json_str)
            .map_err(|e| InspectorError::Parse(format!("Failed to parse JSON: {}", e)))?;

        Ok(entries)
    }

    /// Parse inventory.img (XML format)
    fn parse_inventory(path: &Path) -> Result<Inventory> {
        let content = fs::read_to_string(path)?;

        let inventory: Inventory = quick_xml::de::from_str(&content)
            .map_err(|e| InspectorError::Parse(format!("Failed to parse inventory XML: {}", e)))?;

        Ok(inventory)
    }

    /// Parse fds.img (XML format)
    fn parse_fds(path: &Path) -> Result<Vec<FileDescriptor>> {
        let content = fs::read_to_string(path)?;

        // FDS XML format: <fd id="..." fd="..." type="..." ... />
        // We'll parse it manually for more flexibility
        let mut fds = Vec::new();
        let reader = &mut content.as_bytes();

        let mut parser = quick_xml::Reader::from_reader(reader);
        parser.config_mut().trim_text(true);

        let mut buf = Vec::new();

        loop {
            match parser.read_event_into(&mut buf) {
                Ok(quick_xml::events::Event::Start(e)) => {
                    if e.name().as_ref() == b"fd" {
                        let mut fd_entry = FileDescriptor {
                            fd: 0,
                            fd_type: String::new(),
                            path: None,
                            flags: None,
                            id: None,
                            mnt_id: None,
                        };

                        for attr in e.attributes() {
                            if let Ok(attr) = attr {
                                let key = std::str::from_utf8(attr.key.as_ref())
                                    .unwrap_or("");
                                let value = std::str::from_utf8(&attr.value)
                                    .unwrap_or("");

                                match key {
                                    "fd" => fd_entry.fd = value.parse().unwrap_or(0),
                                    "type" => fd_entry.fd_type = value.to_string(),
                                    "path" => fd_entry.path = Some(value.to_string()),
                                    "flags" => fd_entry.flags = Some(value.to_string()),
                                    "id" => fd_entry.id = Some(value.to_string()),
                                    "mnt_id" => fd_entry.mnt_id = value.parse().ok(),
                                    _ => {}
                                }
                            }
                        }

                        fds.push(fd_entry);
                    }
                }
                Ok(quick_xml::events::Event::Eof) => break,
                Err(e) => return Err(InspectorError::Parse(format!("XML parse error: {}", e))),
                _ => {}
            }
            buf.clear();
        }

        Ok(fds)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::Write;
    use tempfile::TempDir;

    #[test]
    fn test_parse_pstree_json() {
        let json_data = r#"[{
            "pid": 1,
            "ppid": 0,
            "pgid": 1,
            "sid": 1,
            "comm": "init",
            "nthreads": 1
        }, {
            "pid": 2,
            "ppid": 1,
            "pgid": 1,
            "sid": 1,
            "comm": "sh",
            "nthreads": 1
        }]"#;

        let mut binary_data = vec![0u8; 100]; // Mock binary header
        binary_data.extend_from_slice(json_data.as_bytes());

        let tmp = TempDir::new().unwrap();
        let pstree_path = tmp.path().join("pstree.img");
        fs::write(&pstree_path, binary_data).unwrap();

        let result = CriuParser::parse_pstree(&pstree_path).unwrap();

        assert_eq!(result.len(), 2);
        assert_eq!(result[0].pid, 1);
        assert_eq!(result[0].comm, "init");
        assert_eq!(result[1].pid, 2);
        assert_eq!(result[1].ppid, 1);
    }

    #[test]
    fn test_parse_inventory_xml() {
        let xml_data = r#"<?xml version="1.0" encoding="UTF-8"?>
<inventory>
    <magic>CRIUIMG</magic>
    <arch>x86_64</arch>
    <pages_id>12345</pages_id>
    <gen_id>67890</gen_id>
    <entries>42</entries>
    <image_pages>1024</image_pages>
</inventory>"#;

        let tmp = TempDir::new().unwrap();
        let inventory_path = tmp.path().join("inventory.img");
        fs::write(&inventory_path, xml_data).unwrap();

        let result = CriuParser::parse_inventory(&inventory_path).unwrap();

        assert_eq!(result.magic, "CRIUIMG");
        assert_eq!(result.arch, "x86_64");
        assert_eq!(result.pages_id, "12345");
        assert_eq!(result.entries, 42);
        assert_eq!(result.image_pages, Some(1024));
    }

    #[test]
    fn test_parse_fds_xml() {
        let xml_data = r#"<?xml version="1.0" encoding="UTF-8"?>
<fds>
    <fd id="1" fd="0" type="pipe" flags="02000000" />
    <fd id="2" fd="1" type="file" path="/etc/hosts" flags="0100000" />
    <fd id="3" fd="2" type="socket" flags="02000002" />
</fds>"#;

        let tmp = TempDir::new().unwrap();
        let fds_path = tmp.path().join("fds.img");
        fs::write(&fds_path, xml_data).unwrap();

        let result = CriuParser::parse_fds(&fds_path).unwrap();

        assert_eq!(result.len(), 3);
        assert_eq!(result[0].fd, 0);
        assert_eq!(result[0].fd_type, "pipe");
        assert_eq!(result[1].fd, 1);
        assert_eq!(result[1].path, Some("/etc/hosts".to_string()));
        assert_eq!(result[2].fd_type, "socket");
    }
}
